compiler/src/iree/compiler/Codegen/Common/ResolveWorkgroupCountHints.cpp

```diff
@@ -0,0 +1,1043 @@
+      callSlice.callee.getArguments();
+  SmallVector<BlockArgument> argsToUnmap;
+  auto pushOperandSlice = [&](OperandSlice &operandSlice) {
+    if (failed(materializeSlice(rewriter, map, workloadVals, operandSlice.slice,
+                                operandSlice.ordinalLeaves))) {
+      return failure();
+    }
+    // Link the cloned slice for this operand to the basic block arg of the
+    // callee. The function arg is added to `argsToUnmap` so that we can remove
+    // the mapping after processing this callsite (and risking stale mappings).
+    Value curr = map.lookup(operandSlice.operand->get());
+    BlockArgument bbArg = funcArgs[operandSlice.operand->getOperandNumber()];
+    map.map(bbArg, curr);
+    argsToUnmap.push_back(bbArg);
+    return success();
+  };
+
+  for (OperandSlice &operandSlice : callSlice.requiredOperands) {
+    assert(llvm::all_of(operandSlice.funcArgLeaves,
+                        [&](BlockArgument b) { return map.contains(b); }) &&
+           "unexpected unmapped required operand");
+    if (failed(pushOperandSlice(operandSlice))) {
+      callSlice.callee->emitOpError(
+          "failed to materialize operand slice for function call.");
+      return failure();
+    }
+  }
+
+  for (OperandSlice &operandSlice : callSlice.optionalOperands) {
+    if (!llvm::all_of(operandSlice.funcArgLeaves,
+                      [&](BlockArgument b) { return map.contains(b); })) {
+      continue;
+    }
+    if (failed(pushOperandSlice(operandSlice))) {
+      return failure();
+    }
+  }
+
+  FailureOr<SmallVector<Value, 3>> sizes = materializeFunc(
+      rewriter, map, workloadVals, sliceMap[callSlice.callee], sliceMap);
+  if (failed(sizes)) {
+    return failure();
+  }
+
+  // Erase the arg mapping for the function args. Each call to a function
+  // provides a different mapping and may have a different set of optional
+  // args.
+  for (Value v : argsToUnmap) {
+    map.erase(v);
+  }
+
+  assert(sizes->size() == 3 &&
+         "function materialization produced too few values");
+
+  if (placeholder) {
+    Operation *parent = placeholder->getParentOp();
+    SmallVector<Value, 3> newSizes = parent->getResults();
+    rewriter.replaceOp(placeholder, *sizes);
+    *sizes = newSizes;
+    rewriter.setInsertionPointAfter(parent);
+  }
+
+  return *sizes;
+}
+
+/// Materializes the IR for all contained hint and required call ops at the
+/// current insertion point and then takes the maximum values of each
+/// elementwise. For example:
+///
+/// ```
+/// func.func @entry_point() {
+///   iree_codegen.workgroup_count_hint %x, %y, %z
+///   iree_codegen.workgroup_count_hint %a, %b, %c
+///   func.call @constant_hint()
+/// }
+/// func.func @constant_hint() {
+///   iree_codegen.workgroup_count_hint 1, 2, 3
+/// }
+/// ```
+///
+/// This would resolve to:
+///
+/// ```
+/// hal.executable.export @entry_point count(%x, %y, %z, %a, %b, %c) {
+///   %wx = max(%x, %a, 1)
```

Comment:
I am not sure this makes sense to me. Having max along each individual dimension seems very strange and very hard to reason about. The only logical thing I can think of is to linearize and then take the max. Everything else seems to be hard to navigate.